force-app/main/default/classes/InstallationNotifications.cls

```diff
@@ -0,0 +1,113 @@
+      its contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2021
+* @description Domain for building and publishing install script triggered user notifications.
+*/
+
+public inherited sharing class InstallationNotifications {
+
+   private List<Custom_Notification__mdt> activeNotifications;
+   private List<CustomNotificationType> customNotificationTypes;
+   private Map<String, CustomNotificationType> activeNotificationsByType;
+
+   public InstallationNotifications() {
+      this.customNotificationTypes = notificationTypeSelector.getCustomNotificationTypes();
+      this.activeNotifications = installationNotificationsSelector.getActiveCustomNotifications();
+      this.activeNotificationsByType = buildActiveNotificationsByType(this.customNotificationTypes);
+   }
+
+   @TestVisible
+   private InstallationNotificationsSelector installationNotificationsSelector {
+      get {
+         if (installationNotificationsSelector == null) {
+            installationNotificationsSelector = new InstallationNotificationsSelector();
+         }
+         return installationNotificationsSelector;
+      }
+      set;
+   }
+
+   @TestVisible
+   private CustomNotificationTypeSelector notificationTypeSelector {
+      get {
+         if (notificationTypeSelector == null) {
+            notificationTypeSelector = new CustomNotificationTypeSelector();
+         }
+         return notificationTypeSelector;
+      }
+      set;
+   }
+
+   private Map<String, CustomNotificationType> buildActiveNotificationsByType (
+       List<CustomNotificationType> notifications) {
+      Map<String, CustomNotificationType> customNotificationByType = new Map<String, CustomNotificationType>();
+      for (CustomNotificationType customNotificationType : notifications) {
+         customNotificationByType.put(customNotificationType.CustomNotifTypeName, customNotificationType);
+      }
+      return customNotificationByType;
+   }
+
+   public void publish() {
+      String DUMMY_ID = '000000000000000AAA';
```

Comment:
Is this dummy id necessary?  If so, a quick comment to state why will be helpful for posterity.

Reply:
Yes, it's important since we're not doing the redirection in this work item and a target Id/page reference is required. It will be replaced in the next WI and also it's advised in the [documentation](https://developer.salesforce.com/docs/atlas.en-us.apexref.meta/apexref/apex_class_Messaging_CustomNotification.htm) to handle notifications without a target id this way.